ledger/participant-integration-api/src/main/scala/platform/index/BuffersUpdater.scala

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.index
+
+import java.util.concurrent.atomic.AtomicReference
+
+import akka.stream.scaladsl.{Keep, RestartSource, Sink, Source}
+import akka.stream._
+import akka.{Done, NotUsed}
+import com.daml.ledger.participant.state.v1.Offset
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.index.BuffersUpdater._
+import com.daml.platform.store.appendonlydao.EventSequentialId
+import com.daml.platform.store.appendonlydao.events.{Contract, Key, Party}
+import com.daml.platform.store.dao.events.ContractStateEvent
+import com.daml.platform.store.interfaces.TransactionLogUpdate
+import com.daml.scalautil.Statement.discard
+
+import scala.concurrent.duration.{DurationInt, FiniteDuration}
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.{Failure, Success}
+import scala.util.control.NonFatal
+
+/** Creates and manages a subscription to a transaction log updates source
+  * (see [[com.daml.platform.store.dao.LedgerDaoTransactionsReader.getTransactionLogUpdates]]
+  * and uses for updating:
+  *    * The transactions buffer used for in-memory Ledger API serving.
+  *    * The mutable contract state cache.
+  *
+  * @param subscribeToTransactionLogUpdates Subscribe to the transaction log updates stream starting at a specific
+  *                                         `(Offset, EventSequentialId)`.
+  * @param updateTransactionsBuffer Trigger externally the update of the transactions buffer.
+  * @param toContractStateEvents Converts [[TransactionLogUpdate]]s to [[ContractStateEvent]]s.
+  * @param updateMutableCache Trigger externally the update of the mutable contract state cache.
+  *                           (see [[com.daml.platform.store.cache.MutableCacheBackedContractStore]])
+  * @param minBackoffStreamRestart Minimum back-off before restarting the transaction log updates stream.
+  * @param mat The Akka materializer.
+  * @param loggingContext The logging context.
+  * @param executionContext The execution context.
+  */
+private[index] class BuffersUpdater(
+    subscribeToTransactionLogUpdates: SubscribeToTransactionLogUpdates,
+    updateTransactionsBuffer: (Offset, TransactionLogUpdate) => Unit,
+    toContractStateEvents: TransactionLogUpdate => Iterator[ContractStateEvent],
+    updateMutableCache: ContractStateEvent => Unit,
```

Comment:
@tudor-da : is there a particular reason that this `BufferUpdater` takes these three individual elements instead of just a `updateBuffers :: (Offset, TransactionLogUpdate) => Unit` argument?

In the latter case, the caller could take care of doing what it needs doing for each `TransactionLogUpdate`; i.e., we'd gain more context independence.

Reply:
It might well be that this file is just a glue code file, and it makes sense to have the methods in here. Even in that case though you might consider factoring out the pure code that updates the multiple caches, as this way the reader knows that these calls always happen together in the same way.

Reply:
`updateBuffers :: (Offset, TransactionLogUpdate) => Unit` is the actual essence of this class. Factoring it out fully would leave this class an empty shell.

Reply:
> It might well be that this file is just a glue code file, and it makes sense to have the methods in here. Even in that case though you might consider factoring out the pure code that updates the multiple caches, as this way the reader knows that these calls always happen together in the same way.

Sensible suggestion, thanks. I have extracted the logic that updates the buffers from `BuffersUpdater` and passed it as a parameter. That logic is assembled in the `BuffersUpdater$` companion object.

Reply:
:+1: